libsolidity/ast/ASTJsonImporter.cpp

```diff
@@ -0,0 +1,881 @@
+	vector<pair<ASTPointer<Identifier>, ASTPointer<ASTString>>> symbolAliases;
+	for (auto& tuple: _node["symbolAliases"])
+	{
+		symbolAliases.push_back( make_pair(
+			createIdentifier(tuple["foreign"]),
+			tuple["local"].isNull() ? nullptr : make_shared<ASTString>(tuple["local"].asString())
+		));
+	}
+	ASTPointer<ImportDirective> tmp = createASTNode<ImportDirective>(
+		_node,
+		path,
+		unitAlias,
+		move(symbolAliases)
+	);
+	tmp->annotation().absolutePath = _node["absolutePath"].asString();
+	return tmp;
+}
+
+ASTPointer<ContractDefinition> ASTJsonImporter::createContractDefinition(Json::Value const& _node){
+	std::vector<ASTPointer<InheritanceSpecifier>> baseContracts;
+	for (auto& base: _node["baseContracts"])
+		baseContracts.push_back(createInheritanceSpecifier(base));
+	std::vector<ASTPointer<ASTNode>> subNodes;
+	for (auto& subnode: _node["nodes"])
+		subNodes.push_back(convertJsonToASTNode(subnode));
+	return createASTNode<ContractDefinition>(
+		_node,
+		make_shared<ASTString>(_node["name"].asString()),
+		nullOrASTString(_node, "documentation"),
+		baseContracts,
+		subNodes,
+		contractKind(_node)
+	);
+}
+
+ASTPointer<InheritanceSpecifier> ASTJsonImporter::createInheritanceSpecifier(Json::Value const& _node)
+{
+	std::vector<ASTPointer<Expression>> arguments;
+	for (auto& arg: _node["arguments"])
+		arguments.push_back(castPointer<Expression>(convertJsonToASTNode(arg)));
+	return createASTNode<InheritanceSpecifier>(
+		_node,
+		createUserDefinedTypeName(_node["baseName"]),
+		arguments.size() == 0 ? nullptr : make_unique<std::vector<ASTPointer<Expression>>>(arguments)
+	);
+}
+
+ASTPointer<UsingForDirective> ASTJsonImporter::createUsingForDirective(Json::Value const& _node)
+{
+	return createASTNode<UsingForDirective>(
+		_node,
+		createUserDefinedTypeName(member(_node, "libraryName")),
+		!_node["typeName"].isNull() ? castPointer<TypeName>(convertJsonToASTNode(_node["typeName"])) : nullptr
+	);
+}
+
+ASTPointer<ASTNode> ASTJsonImporter::createStructDefinition(Json::Value const& _node)
+{
+	std::vector<ASTPointer<VariableDeclaration>> members;
+	for (auto& member: _node["members"])
+		members.push_back(createVariableDeclaration(member));
+	return createASTNode<StructDefinition>(
+		_node,
+		memberAsASTString(_node, "name"),
+		members
+	);
+}
+
+ASTPointer<EnumDefinition> ASTJsonImporter::createEnumDefinition(Json::Value const& _node)
+{
+	std::vector<ASTPointer<EnumValue>> members;
+	for (auto& member: _node["members"])
+		members.push_back(createEnumValue(member));
+	return createASTNode<EnumDefinition>(
+		_node,
+		memberAsASTString(_node, "name"),
+		members
+	);
+}
+
+ASTPointer<EnumValue> ASTJsonImporter::createEnumValue(Json::Value const& _node)
+{
+	return createASTNode<EnumValue>(
+		_node,
+		memberAsASTString(_node, "name")
+	);
+}
+
+ASTPointer<ParameterList> ASTJsonImporter::createParameterList(Json::Value const&  _node)
+{
+	std::vector<ASTPointer<VariableDeclaration>> parameters;
```

Comment:
Do you think it is possible to create a template function such that you could just call
```
createVector(_node["parameters"], createVariableDeclaration)
```
?
I think for this to really work, `createVariableDeclaration` needs to be a non-member function. But as far as I can see, most of your functions do not use the state of the ASTJsonImporter anyway, right?

Reply:
> most of your functions do not use the state of the ASTJsonImporter anyway, right?

almost true. 
Most members are used only in ```jsonToSourceUnit()```, which is the entryPoint for the class anyways. 

```createVariableDeclaration```  and the other ```createX```()s mostly call eachother, passing around json-nodes, so that would be fine i guess. 
But they all do use the state of the Importer once, namely when they are creating the sourceLocation of their nodes: 
- => they are calling createASTNode(), which in turn call createSourceLocation (which uses  ```m_sourceLocations.size()``` & ```m_currentSourceName```:


```
SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _node)
{
	astAssert(!member(_node, "src").isNull(), "'src' can not be null");
	string srcString = _node["src"].asString();
	vector<string> pos;
	boost::algorithm::split(pos, srcString, boost::is_any_of(":"));
	if (pos.size() != 3 || int(m_sourceLocations.size()) < stoi(pos[2]))
		BOOST_THROW_EXCEPTION(langutil::InvalidAstError() << errinfo_comment("'src'-field ill-formatted or src-index too high"));
	int start = stoi(pos[0]);
	int end = start + stoi(pos[1]);
	// ASSUMPTION: only the name of source is used from here on, the m_source of the CharStream-Object can be empty
	std::shared_ptr<langutil::CharStream> source = make_shared<langutil::CharStream>("", m_currentSourceName);
	return SourceLocation{ start, end,source};
```

One way to avoid that could be by passing those members along as arguments in the createX()s,  but that seems rather inelegant to me. Do you see a better way to do it?